### PR TITLE
<WIP FOR FRANK> Jess/Rinn Liu

### PR DESCRIPTION
**Pull request by juyuanliu**
HI FRANK

### CODE REVIEW

File: src/closure.js [hunk 1]
```diff
@@ -1,21 +1,32 @@
...
+    increment: () => ++count,
+    decrement: () => --count,
+  };
 };
 
 const limitFunctionCallCount = (cb, n) => {
-  // Should return a function that invokes `cb`.
-  // The returned function should only allow `cb` to be invoked `n` times.
+  let count = 0;
+  return (...args) => {
```
Comment (frankfaustino) [inline src/closure.js:23]: Good stuff! 👍

File: src/arrays.js [hunk 2]
```diff
@@ -1,44 +1,71 @@
...
 };
 
 /* STRETCH PROBLEM */
-
 const flatten = (elements) => {
-  // Flattens a nested array (the nesting can be to any depth).
-  // Example: flatten([1, [2], [3, [[4]]]]); => [1, 2, 3, 4];
+  let outPut = [];
+  each(elements, (num) => {
+    if (Array.isArray(num)) outPut = outPut.concat(flatten(num));
```
Comment (frankfaustino) [inline src/arrays.js:63]: Awesome job, Rinn! :D

File: src/arrays.js [hunk 3]
```diff
@@ -1,44 +1,71 @@
+// SAVE ME BABY JESUS FROM THE LINT GODS.
```
Comment (frankfaustino) [inline src/arrays.js:1]: 🤣

File: src/callbacks.js [hunk 4]
```diff
@@ -1,36 +1,40 @@
...
 
 /* STRETCH PROBLEM */
-
 const removeDuplicates = (array, cb) => {
-  // removeDuplicates removes all duplicate values from the given array.
-  // Pass the duplicate free array to the callback function.
-  // Do not mutate the original array.
+  const newArray = array.slice();
+
+  for (let i = 0; i < newArray.length - 1; i++) {
```
Comment (frankfaustino) [inline src/callbacks.js:31]: Your solution isn't actually removing the duplicate item in the array. (The test is only checking if `array.length` was reduced by 1. We need to add another test to check if the duplicate item was removed)

File: src/closure.js [hunk 5]
```diff
@@ -27,6 +38,8 @@ const cacheFunction = (cb) => {
   // If the returned function is invoked with arguments that it has already seen
   // then it should return the cached result and not invoke `cb` again.
   // `cb` should only ever be invoked once for a given set of arguments.
+
+  // TODO: Come back to if sanity intact.
```
Comment (frankfaustino) [inline src/closure.js:42]: If you want to go over this one with me later, let me know!

File: src/objects.js [hunk 6]
```diff
@@ -1,26 +1,29 @@
+// Completed everything except for stretch. Will come back to later!
+
 // Complete the following underscore functions.
 // Reference http://underscorejs.org/ for examples.
 
 const keys = (obj) => {
-  // Retrieve all the names of the object's properties.
-  // Return the keys as strings in an array.
-  // Based on http://underscorejs.org/#keys
+  return Object.getOwnPropertyNames(obj);
```
Comment (frankfaustino) [inline src/objects.js:7]: Nice! `Object.keys()` would work here. One thing you should note, if you use `Object.getOwnPropertyNames()` on an array, you'll get its `length` as well

File: src/objects.js [hunk 7]
```diff
@@ -1,26 +1,29 @@
...
 };
 
 const values = (obj) => {
-  // Return all of the values of the object's own properties.
-  // Ignore functions
-  // http://underscorejs.org/#values
+  return Object.values(obj);
 };
 
 const mapObject = (obj, cb) => {
```
Comment (frankfaustino) [inline src/objects.js:14]: Since this is supposed to emulate `.map()`, we should be creating a new Object and assigning the new values to it.

**Comment by frankfaustino** [comment]
Good stuff, Rinn! 👌 If you want to go over the exercises you didn't complete, just DM me! Keep up the great work
